azurelinuxagent/ga/cgroupapi.py

```diff
@@ -470,55 +476,227 @@ def _get_root_cgroup_path():
+        """
+        Cgroup version specific. Returns True if the cgroup is in the expected slice, False otherwise.
+
+        :param expected_slice: The slice the cgroup is expected to be in.
+        """
+        raise NotImplementedError()
+
+    def get_controller_metrics(self, expected_relative_path=None):
+        """
+        Cgroup version specific. Returns a list of the metrics for the agent supported controllers which are
+        mounted/enabled for the cgroup.
+
+        :param expected_relative_path: The expected relative path of the cgroup. If provided, only metrics for controllers at this expected path will be returned.
+        """
+        raise NotImplementedError()
+
+    def get_processes(self):
+        """
+        Cgroup version specific. Returns a list of all the process ids in the cgroup.
+        """
+        raise NotImplementedError()
+
+
+class CgroupV1(Cgroup):
+    CPU_CONTROLLER = "cpu,cpuacct"
+
+    def __init__(self, cgroup_name, controller_mountpoints, controller_paths):
+        """
+        :param cgroup_name: The name of the cgroup. Used for logging/tracking purposes.
+        :param controller_mountpoints: A dictionary of controller-mountpoint mappings for each agent supported controller which is mounted.
+        :param controller_paths: A dictionary of controller-path mappings for each agent supported controller which is mounted. The path represents the absolute path of the controller.
+        """
+        super(CgroupV1, self).__init__(cgroup_name=cgroup_name)
+        self._controller_mountpoints = controller_mountpoints
+        self._controller_paths = controller_paths
+
+    @staticmethod
+    def get_supported_controllers():
+        return [CgroupV1.CPU_CONTROLLER, CgroupV1.MEMORY_CONTROLLER]
+
+    def check_in_expected_slice(self, expected_slice):
+        in_expected_slice = True
+        for controller, path in self._controller_paths.items():
+            if expected_slice not in path:
+                log_cgroup_warning("The {0} controller for the {1} cgroup is not mounted in the expected slice. Expected slice: {2}. Actual controller path: {3}".format(controller, self._cgroup_name, expected_slice, path), send_event=False)
+                in_expected_slice = False
+
+        return in_expected_slice
+
+    def get_controller_metrics(self, expected_relative_path=None):
+        metrics = []
+
+        for controller in self.get_supported_controllers():
+            controller_metrics = None
+            controller_path = self._controller_paths.get(controller)
+
+            if controller_path is None:
+                log_cgroup_warning("{0} is not mounted for the {1} cgroup; will not track metrics".format(controller, self._cgroup_name), send_event=False)
+                continue
+
+            if expected_relative_path is not None:
+                expected_path = ""
+                controller_mountpoint = self._controller_mountpoints[controller]
```

Comment:
this may fail if controller not found in mountpoint. Consider switch to get and handle None case and log the warning if it;s none

Reply:
My reasoning here was that if the controller exists in self._controller_paths, then it must be mounted. But I will add none check here to be cautious